src/amm.rs

```diff
@@ -34,25 +35,36 @@ macro_rules! to_u128 {
     };
 }
 
+#[derive(PartialEq)]
+#[derive(Debug)]
+pub enum MathError {
+    ZeroInReserve,
+    Overflow,
+    InsufficientOutReserve,
+}
+
 /// Calculating spot price given reserve of selling asset and reserve of buying asset.
 /// Formula : BUY_RESERVE * AMOUNT / SELL_RESERVE
 ///
 /// - `sell_reserve` - reserve amount of selling asset
 /// - `buy_reserve` - reserve amount of buying asset
 /// - `amount` - amount
 ///
-/// Returns None in case of error
-pub fn calculate_spot_price(sell_reserve: Balance, buy_reserve: Balance, amount: Balance) -> Option<Balance> {
-    ensure!(sell_reserve != 0);
+/// Returns MathError in case of error
+pub fn calculate_spot_price(sell_reserve: Balance, buy_reserve: Balance, amount: Balance) -> Result<Balance, MathError> {
+    ensure!(sell_reserve != 0, ZeroInReserve);
+
+    if amount == 0 || buy_reserve == 0 {
+        return Ok(0u128);
```

Comment:
Use the to_u128(0) like it is done elsewhere, so it is consistent and might be easier to refactor ( eg if we decide to change Balance to other type ).